### DISCUSSION TITLE
NiceGUI style navigation bar

### DISCUSSION THREAD

**Discussion by frankhuurman** [zauberzeug/nicegui#1715] [Ideas / Feature Requests]
I tried diving into the repository parts of the NiceGUI website itself to figure out how to create the great looking responsive top navigation bar(so far I've been using ui.rows, tabs, tab_panels and menus to get some navigation going but it feels a bit awkward to handle it like that) but I couldn't figure out where the navbar is built or if everything is html + styling.

Would it possible to add a NiceGUI-style sticky navbar component that can handle a .svg file for a logo and link to a few pages like the website itself? Or is this something too complex to be made into a native NiceGUI component?

To take the NiceGUI website as an example, it would look something like:
```python
navbar = ui.navbar(bgcolor="#5898d4", logo="images/mylogo.svg", logo_align="left", color="#fff", blur=5)
with navbar:
    ui.nav(text="Installation", target="/#installation")
    ui.nav(text="Features", target="/#features")
    ui.nav(text="Demos", target="/#demos")
    ui.nav(text="Documentation", target="/documentation")
    ui.nav(text="Examples", target="/#examples")
    ui.nav(text="Why?", target="/#why")
    ui.nav(icon="images/search.svg", target="/search or search()")  # Unsure if target should also accept a callable instead of a page
    ui.nav(icon="images/nightmode.svg", target="/nightmode or nightmode()")
    ui.nav(icon="images/discord.svg", target="https://discord.gg/TEpFeAaF4f")
    ui.nav(icon="images/reddit.svg", target="https://www.reddit.com/r/nicegui/")
    ui.nav(icon="images/github.svg", target="https://github.com/zauberzeug/nicegui/")
    ui.nav(icon="images/star-animation.svg", target="https://github.com/zauberzeug/nicegui/")
```

Of course this doesn't take into account the column spacings that are currently between the NiceGUI logo, the text navigation items, the search icon navigation item and the other icon navigation items but that might be a bit too much customization for a simple sticky navbar. 

Hope it's a good idea!

**Comment by frankhuurman** [comment]
ah, as much as I've read through the documentation already, I missed the https://nicegui.io/documentation#page_layout that will probably help with building layouts like the NiceGUI one.

**Comment by falkoschindler** [comment]
Hi @frankhuurman, the [page layout](https://nicegui.io/documentation#page_layout) helps indeed to build a custom page header. But we still find your feature request very valid. Users keep requesting simpler ways to quickly generate page layouts, e.g. using a navigation bar.

Your example code looks very similar to how you would use `ui.header` and `ui.link`. So I'm not sure if this would justify new UI elements `ui.navbar` and `ui.nav`.

But I could imagine a more highlevel `ui.navbar` for those who don't want to deal with any details, but just want to get some navigation working:
```py
ui.navbar(
    title='My Virtual Cookbook',
    logo='logo.png',
    links={
        'Home': '/',
        'Recipes': '/recipes',
        'Shopping List': '/shopping-list',
        'About': '/about',
    },
    background_color='#f5f5f5',
    breakpoint=450,
)
```
- The `logo` parameter could accept routes or local paths (similar to `ui.image`).
- The `links` dictionary maps link titles to routes. But we could also accept callables or page generator functions.
- The `breakpoint` parameter might be useful to fine-tune when to collapse the links to a `ui.menu`.
- We could even call it `ui.navigation` and add a parameter `position: Literal['header', 'left', 'right'] = 'header'` to allow moving it into a sidebar.

All in all it could be a very handy UI element.

**Reply by frankhuurman** [reply]
Thanks for the reply and your suggestion looks even better :+1: 

My idea was indeed more like "let's just get a quick navbar going for prototyping" without too much customization.
Maybe my suggestion of having the links(and not just the navbar) accept logo parameters similar to ui.image as well would already be a bit too much customization.
Also would rename the links parameter to pages since we also use ui.page to make a new endpoint :smile: 

But this would be the perfect all-inclusive navbar(including the breakpoint to collapse links to a ui.menu) for almost everyone's use-cases. If someone would want more control over it there's always ui.header and ui.link to get it more to their liking.

Can't wait to try something like it in a future release!


_Edit: Plus, this element could be used as a bit of NiceGUI branding if you don't provide any styling parameters. 
Just like Streamlit apps are instantly recognizable, if someone would use ui.navbar and just provide a dict or list of pages it would end up with the same #75aadb blue background color and position of pages/links like the NiceGUI website which would make it recognizable as a NiceGUI site._

**Comment by frankhuurman** [comment]
@falkoschindler 

A bit off-topic but could you help me get some insight on getting the header spacing right?
I'm not that well versed in front-end, Tailwind or Quasar(never even heard of Quasar until I started using NiceGUI) so I might be asking something really basic here.

I've tried all the different justify/space-x/pl-1 tailwind classes, ui.columns and ui.rows with additional css padding/margin and QHeader layout builder things but I can't seem to recreate the NiceGUI header spacing or something close to it. The closest I've gotten was due to chucking in a couple of ui.label("") elements to fill up some space which naturally makes the layout garbled on mobile since it still renders it as an empty object that it needs to take into account with the justify-between spacing.

This code(almost exactly like the page layout example from the NiceGUI docs):
```python
with ui.header(elevated=True).style('background-color: rgb(85, 214, 244)').classes('items-center justify-between'):
        ui.avatar('favorite_border', text_color='grey-11', square=False)

        ui.button("Shop", icon="shopping_cart").props('flat color=blue-grey-7')
        ui.button("Blog", icon="feed").props('flat color=blue-grey-7')
        ui.button("Contact", icon="perm_phone_msg").props('flat color=blue-grey-7')

        ui.button(on_click=lambda: right_drawer.toggle(), icon='menu').props('flat color=white')

    with ui.right_drawer(fixed=False, value=False).style('background-color: #ebf1fa').props('bordered') as right_drawer:
        ui.label('RIGHT DRAWER')
    
    ui.label('CONTENT')
    [ui.label(f'Line {i}') for i in range(100)]

    with ui.footer().style('background-color: #3874c8'):
        ui.label('FOOTER')
```

Turns into this on desktop:
![Screenshot from 2023-10-06 17-17-41](https://github.com/zauberzeug/nicegui/assets/19980966/a0795d49-65d5-40d9-a9b8-620eb4b9c8c8)

And into this on mobile:
![image](https://github.com/zauberzeug/nicegui/assets/19980966/f69f5fd0-21ae-4034-8e56-64a419aa5d44)


The spacing on desktop looks weirdly stretched out like this so I'm wondering how to divide the header into a logo part on the left, 3-ish navigation buttons part in the center and the hamburger menu part on the right. As I mentioned, I tried to do this using ui.columns inside of the header but that failed.

Ideally it would look something like this:
![betterspacing](https://github.com/zauberzeug/nicegui/assets/19980966/6d98bfeb-d043-43e6-94dd-24208e0b00a2)

Additionally I've got no clue how to make this layout work on mobile. I could of course duplicate the navigation items to the ui.right_drawer but I've got no idea how to move the original buttons out of sight/hide them once the drawer opens.
Or change the navigation items to just the icons without text to save space and have all the items in the same row.

A bit like this(excuse the poor photoshop):
![mobile-example](https://github.com/zauberzeug/nicegui/assets/19980966/9c5d0c15-acc3-4a07-9d1e-3db6b80b26ae)

But I guess this would make the hamburger menu redundant again.

**Reply by falkoschindler** [reply]
I would
- group the buttons into a `ui.row` so that the "justify-between" doesn't tear them apart,
- use a breakpoint "max-sm:" to hide the button group on small screens, and
- add a copy of the row with icon-only buttons and the breakpoint "sm:" to show them on small screens.

```py
with ui.header().classes('items-center justify-between'):
    ui.avatar('favorite_border')
    with ui.row().classes('max-sm:hidden'):
        ui.button('Shop', icon='shopping_cart').props('flat color=white')
        ui.button('Blog', icon='feed').props('flat color=white')
        ui.button('Contact', icon='perm_phone_msg').props('flat color=white')
    with ui.row().classes('sm:hidden'):
        ui.button(icon='shopping_cart').props('flat color=white')
        ui.button(icon='feed').props('flat color=white')
        ui.button(icon='perm_phone_msg').props('flat color=white')
    ui.button(icon='menu').props('flat color=white')
```

<img width="1512" alt="Screenshot 2023-10-06 at 19 16 21" src="https://github.com/zauberzeug/nicegui/assets/5767091/f8b323aa-e6d6-453d-8f92-6b5a859df103">
<img width="582" alt="Screenshot 2023-10-06 at 19 16 29" src="https://github.com/zauberzeug/nicegui/assets/5767091/18f74aab-7840-4db3-b325-63f6abcb99c4">

**Reply by frankhuurman** [reply]
Ah! exactly what I needed! Thanks so much :smile:
It wasn't an obvious action to me to add only the rows inside of the header(my assumption was that the `ui.avatar` in the header itself would not be seen as part of the same row as the 2 new ones and therefore would mess with the alignments), that's why I tried it with columns to keep the horizontal alignment intact.

But amazing! Now I can implement the webshop logic I've built the past few months and I've got myself a fully fledged NiceGUI webshop :+1: :tada: 

I'm such a big fan of the framework and it keeps on getting better with every release :D
